0015-authority.md

```diff
@@ -0,0 +1,478 @@
+public class SampleClass : NetworkBehaviour{
+    
+    [ServerRpc]
+    public void SetPositionServerRpc(Vector3 position)
+    {
+        SetPosition(position)
+    }
+
+    public void SetPosition(Vector3 position)
+    {
+        transform.Position = position;
+    }
+
+    public void Update()
+    {
+        if(HasAuthority)
+        {
+            SetPosition(m_TargetPosition);
+        }
+        else
+        {
+            SetPositionServerRpc(m_TargetPosition);
+        }
+    }
+}
+```
+
+## NetworkVariables
+NetworkVariables will get an additional write and read permission for the authority.
+
+`NetworkVariablePermission.Authority` can be used to make a NetworkVariable only writable/readable by the authority.
+This can be used to write code which works for any authority mode. For instance like this.
+```csharp
+public class SampleClass : NetworkBehaviour{
+    
+    var m_MyInput = new NetworkVariable<bool>(
+        new NetworkVariableSettings()
+        {
+            ReadPermission = NetworkVariablePermission.Authority,
+            WritePermission = NetworkVariablePermission.OwnerOnly
+        });
+
+    public void Update()
+    {
+        // Owner client gives inputs into network variable.
+        if (IsOwner)
+        {
+            m_MyInput.Value = Input.GetKeyDown(KeyCode.Space);
+        }
+
+        // The authority which could either be the owner or the server based on the authority mode uses the input to move.
+        if(HasAuthority)
+        {
+            if(m_MyInput.Value)
+            {
+                m_Rigidbody.AddForce(transform.up * m_Thrust);
+            }
+        }
+    }
+}
+```
+*The authority system allows NetworkVariables to be flexibly used as inputs.*
+
+## Changing Authority
+
+The `Authority` mode of a `NetworkObject` can be changed, on the server only, by calling `networkObject.ChangeAuthority(Authority.Server);`.
+
+This means which peer has authority also changes if the ownership of an object changes when in `owner` authority mode. This means the `HasAuthority` property should never be cached but always accessed directly.
+
+## OnGainedAuthority / OnLostAuthority
+
+`OnGainedAuthority` and `OnLostAuthority` functions can be overridden to get informed about changes to the `HasAuthority` status. `OnGainedAuthority` gets called when `HasAuthority` switches from false to true. `OnLostAuthority` gets called when it switches to false from true
+```csharp
+public override void OnGainedAuthority()
+{
+    Debug.Log("Authority Gained :)");
+}
+
+public override void OnLostAuthority()
+{
+    Debug.Log("Authority Lost :(");
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why is Authority a mode?
+Why is authority not something like ownership which you give to a client. The reason for that is because authority is a relation between multiple peers. In an authoritative model a set of peers pass inputs to the authoritative peer for processing.
```

Comment:
"In an authoritative model a set of peers pass inputs to the authoritative peer for processing."
not with client authority. then each clients would use their own inputs and only send their authoritative state to the server.

Reply:
In my mind with client authority the owner peer passes inputs to itself (the authoritative peer). But yeah I can see why the wording might be confusing.